content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md

```diff
@@ -87,60 +87,65 @@ spec:
+
+          // GetPreferredAllocation returns a preferred set of devices to allocate
+          // from a list of available ones. The resulting preferred allocation is not
+          // guaranteed to be the allocation ultimately performed by the
+          // devicemanager. It is only designed to help the devicemanager make a more
+          // informed allocation decision when possible.
+          rpc GetPreferredAllocation(PreferredAllocationRequest) returns (PreferredAllocationResponse) {}
+
+          // PreStartContainer is called, if indicated by Device Plugin during registeration phase,
+          // before each container start. Device plugin can run device specific operations
+          // such as resetting the device before making devices available to the container.
+          rpc PreStartContainer(PreStartContainerRequest) returns (PreStartContainerResponse) {}
+    }
+    ```
+
+    {{< note >}}
+    Plugins are not required to provide useful implementations for
+    `GetPreferredAllocation()` or `PreStartContainer()`. Flags indicating which
+    (if any) of these calls are available should be set in the `DevicePluginOptions`
+    message sent back by a call to `GetDevicePluginOptions()`. The `kubelet` will
+    always call `GetDevicePluginOptions()` to see which optional functions are
+    available, before calling any of them directly.
+    {{< /note >}}
+
+1. The plugin registers itself with the kubelet through the Unix socket at host
   path `/var/lib/kubelet/device-plugins/kubelet.sock`.
 
-* After successfully registering itself, the device plugin runs in serving mode, during which it keeps
-  monitoring device health and reports back to the kubelet upon any device state changes.
-  It is also responsible for serving `Allocate` gRPC requests. During `Allocate`, the device plugin may
-  do device-specific preparation; for example, GPU cleanup or QRNG initialization.
-  If the operations succeed, the device plugin returns an `AllocateResponse` that contains container
-  runtime configurations for accessing the allocated devices. The kubelet passes this information
-  to the container runtime.
+    {{< note >}}
+    The ordering of the workflow is important. A plugin MUST start serving gRPC
+    service before registering itself with kubelet for successful registration.
+    {{< /note >}}
+
+1. After successfully registering itself, the device plugin runs in serving mode, during which it keeps
+monitoring device health and reports back to the kubelet upon any device state changes.
+It is also responsible for serving `Allocate` gRPC requests. During `Allocate`, the device plugin may
+do device-specific preparation; for example, GPU cleanup or QRNG initialization.
+If the operations succeed, the device plugin returns an `AllocateResponse` that contains container
+runtime configurations for accessing the allocated devices. The kubelet passes this information
+to the container runtime.
```

Comment:
I'd suggest we indent the who paragraph for readability.

```suggestion
   monitoring device health and reports back to the kubelet upon any device state changes.
   It is also responsible for serving `Allocate` gRPC requests. During `Allocate`, the device plugin may
   do device-specific preparation; for example, GPU cleanup or QRNG initialization.
   If the operations succeed, the device plugin returns an `AllocateResponse` that contains container
   runtime configurations for accessing the allocated devices. The kubelet passes this information
   to the container runtime.
```

Reply:
Done, Thanks!